File: frontend/src/services/activityService.ts
import { activityRepository, syncService } from './repositories';
import type { DailyActivity } from '../types';

export const activityService = {
  async getActivities(patientId: string): Promise<DailyActivity[]> {
    return await activityRepository.getActivities(patientId);
  },

  async toggleActivity(id: string, completed: boolean): Promise<DailyActivity> {
    const updated = await activityRepository.toggleActivityCompletion(id, completed);
    await syncService.queueEvent({
      entityType: 'activity',
      action: 'update',
      payload: updated,
    });
    return updated;
  },

  async createActivity(activity: Omit<DailyActivity, 'id'>): Promise<DailyActivity> {
    const created = await activityRepository.createActivity(activity);
    await syncService.queueEvent({
      entityType: 'activity',
      action: 'create',
      payload: created,
    });
    return created;
  },

  async updateActivity(id: string, data: Partial<DailyActivity>): Promise<DailyActivity> {
    const updated = await activityRepository.updateActivity(id, data);
    await syncService.queueEvent({
      entityType: 'activity',
      action: 'update',
      payload: updated,
    });
    return updated;
  },

  async deleteActivity(id: string): Promise<boolean> {
    const success = await activityRepository.deleteActivity(id);
    await syncService.queueEvent({
      entityType: 'activity',
      action: 'delete',
      payload: { id },
    });
    return success;
  }
};

File: backend/app/services/activity_service.py
from typing import List, Optional
from datetime import date, datetime, timezone
from sqlalchemy.orm import Session
from app.models.activity import DailyActivity
from app.repositories.activity_repo import ActivityRepository
from app.schemas.activity import DailyActivityCreate, DailyActivityUpdate
from app.core.exceptions import NotFoundException


class ActivityService:
    def __init__(self, db: Session):
        self.db = db
        self.activity_repo = ActivityRepository(db)

    def get_activities(
        self, patient_id: str, scheduled_date: Optional[date] = None
    ) -> List[DailyActivity]:
        return self.activity_repo.get_by_patient(patient_id, scheduled_date)

    def toggle_completion(self, activity_id: str, completed: bool) -> DailyActivity:
        act = self.activity_repo.get(activity_id)
        if not act:
            raise NotFoundException("DailyActivity")
        act.completed = completed
        act.completed_at = datetime.now(timezone.utc) if completed else None
        self.db.commit()
        self.db.refresh(act)
        return act

    def create_activity(self, activity_data: DailyActivityCreate) -> DailyActivity:
        return self.activity_repo.create(activity_data)

    def update_activity(self, id: str, update_data: DailyActivityUpdate) -> DailyActivity:
        act = self.activity_repo.get(id)
        if not act:
            raise NotFoundException("DailyActivity")
        return self.activity_repo.update(act, update_data)

    def delete_activity(self, id: str) -> bool:
        return self.activity_repo.delete(id)
